expression/constant_propagation.go

```diff
@@ -138,10 +86,72 @@ func (s *propagateConstantSolver) propagateEQ() {
 	}
 }
 
-// validPropagateCond checks if the cond is an expression like [column op constant] and op is in the funNameMap.
-func (s *propagateConstantSolver) validPropagateCond(cond Expression, funNameMap map[string]bool) (*Column, *Constant) {
+// propagateColumnEQ propagates expressions like 'column A = column B' by adding extra filters
+// 'expression(..., column B, ...)' propagated from 'expression(..., column A, ...)' as long as:
+//
+//  1. The expression is deterministic
+//  2. The expression doesn't have any side effect
+//
+// e.g. For expression a = b and b = c and c = d and c < 1 , we can get extra a < 1 and b < 1 and d < 1.
+// However, for a = b and a < rand(), we cannot propagate a < rand() to b < rand() because rand() is non-deterministic
+//
+// This propagation may bring redundancies that we need to resolve later, for example:
+// for a = b and a < 3 and b < 3, we get new a < 3 and b < 3, which are redundant
+// for a = b and a < 3 and 3 > b, we get new b < 3 and 3 > a, which are redundant
+// for a = b and a < 3 and b < 4, we get new a < 4 and b < 3 but should expect a < 3 and b < 3
+// for a = b and a in (3) and b in (4), we get b in (3) and a in (4) but should expect 'false'
+//
+// TODO: remove redundancies later
```

Comment:
It's better to add a github issue for this TODO item.

Reply:
OK.